protocol/src/main/java/org/jboss/as/protocol/mgmt/AsyncToCompletableFutureAdapter.java

```diff
@@ -0,0 +1,164 @@
+            asyncFuture.addListener(new AsyncFuture.Listener<T, Object>() {
+
+                @Override
+                public void handleComplete(AsyncFuture<? extends T> future, Object attachment) {
+                    try {
+                        Adapter.this.complete(transformer.apply(future.get()));
+                    } catch (InterruptedException | ExecutionException e) {
+                        Adapter.this.completeExceptionally(e);
+                    } finally {
+                        latch.countDown();
+                    }
+                }
+
+                @Override
+                public void handleFailed(AsyncFuture<? extends T> future, Throwable cause, Object attachment) {
+                    try {
+                        Adapter.this.completeExceptionally(cause);
+                    } finally {
+                        latch.countDown();
+                    }
+
+                }
+
+                @Override
+                public void handleCancelled(AsyncFuture<? extends T> future, Object attachment) {
+                    try {
+                        Adapter.this.internalCancel();
+                    } finally {
+                        latch.countDown();
+                    }
+                }
+            }, this);
+        }
+
+        @Override
+        public boolean cancel(boolean mayInterruptIfRunning) {
+
+            // Only try and cancel our asyncFuture once
+            if (!cancellable.compareAndSet(false, true)) {
+                // If another thread is trying to cancel the asyncFuture, await their work
+                awaitLatch();
+                return super.cancel(mayInterruptIfRunning);
+            }
+
+            try {
+                // Cancel our AsyncFuture.
+                // These next two lines mimic the impl of AsyncFutureTask.cancel(boolean), except
+                // instead of the first line being a call to AsyncFutureTask.asyncCancel(boolean)
+                // we allow calling an alternative function passed into our constructor.
+                // Passing in such a function takes the place of some old code that wrapped the
+                // AsyncFuture in a delegating subclass that overrode asyncCancel.
+                asyncCancelConsumer.accept(mayInterruptIfRunning);
+                boolean afCancelled = asyncFuture.awaitUninterruptibly() == AsyncFuture.Status.CANCELLED;
+
+                if (!afCancelled) {
+                    // Our AsyncFuture.Listener completes this future asynchronously,
+                    // so block until it has done so before providing the resulting
+                    // state by calling super.cancel.
+                    //
+                    // We do this because canceling the async future may result in
+                    // the canceled management op returning with either outcome=cancelled
+                    // or outcome=failed. We're in this block, so we know it's the latter.
```

Comment:
> We're in this block, so we know it's the latter

I am missing something here, looking at the possible results returned by `asyncFuture.awaitUninterruptibly()`, one of them is `AsyncFuture.Status.COMPLETE`

Can we not arrive here with the AsyncFuture completed successfully?

Reply:
Right; that comment should be improved.